### PR TITLE
Support referring to attribute relationship with Graql `has`

### PR DESCRIPTION
**Pull request by aelred**
This is still using a temporary syntax until we agree on one.

### CODE REVIEW

**Comment by aelred** [comment]
!rtg - random failure with ai.grakn.test.engine.GraknEngineFailoverIT.whenSubmittingTasksToTwoEngines_TheyComplete

File: docs/pages/documentation/3-querying-with-graql/insert-queries.md [hunk 1]
```diff
@@ -149,7 +149,7 @@ You can also specify a variable to represent the relation connecting the thing a
 <div class="tab-content">
 <div role="tabpanel" class="tab-pane active" id="shell6">
 <pre>
-insert isa person has REIFIED {{ $r }} identifier "Fuchsia Groan";
+insert isa person has 🍌 $r identifier "Fuchsia Groan";
```
Comment (fppt) [inline docs/pages/documentation/3-querying-with-graql/insert-queries.md]: No

**Comment by aelred** [comment]
@kasper-piskorski when you're back, this new feature will need reasoner support!

@fppt time to review. No more bananas I promise

File: docs/pages/documentation/3-querying-with-graql/insert-queries.md [hunk 2]
```diff
@@ -138,9 +138,30 @@ qb.insert(var().isa("person").has("identifier", "Fuchsia Groan").has("gender", "
 </div> <!-- tab-pane -->
 </div> <!-- tab-content -->
 
+You can also specify a variable to represent the relation connecting the thing and the resource:
```
Comment (fppt) [inline docs/pages/documentation/3-querying-with-graql/insert-queries.md]: What's a `resource`?  ;)

File: docs/pages/documentation/3-querying-with-graql/match-queries.md [hunk 3]
```diff
@@ -104,6 +106,27 @@ qb.match(var("x").has("identifier", contains("Bar")));
 </div> <!-- tab-pane -->
 </div> <!-- tab-content -->
 
+You can also specify a variable to represent the relation connecting the thing and the resource:
```
Comment (fppt) [inline docs/pages/documentation/3-querying-with-graql/match-queries.md]: What's a relation? 
What's a resource?

File: grakn-core/src/main/java/ai/grakn/concept/Thing.java [hunk 4]
```diff
@@ -80,6 +80,14 @@
     Thing resource(Resource resource);
 
     /**
+     * Creates a relationship from this instance to the provided resource.
```
Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/concept/Thing.java]: For the sake of my future sanity please to the following:

```
Creates a {@link Relationship} from this {@link Thing} to the provided {@link Attribute}.
```

I know it's more tedious but it makes refactors 100x easier

File: grakn-core/src/main/java/ai/grakn/graql/VarPattern.java [hunk 5]
```diff
@@ -107,21 +107,33 @@
      * the variable must have a resource of the given type that matches the given atom
      *
      * @param type a resource type in the ontology
-     * @param varPattern a variable pattern representing a resource
+     * @param resource a variable pattern representing a resource
```
Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/graql/VarPattern.java]: Same javadoc nitpick

File: grakn-core/src/main/java/ai/grakn/graql/VarPattern.java [hunk 6]
```diff
@@ -107,21 +107,33 @@
...
      * @return this
      */
     @CheckReturnValue
-    VarPattern has(String type, VarPattern varPattern);
+    VarPattern has(String type, VarPattern resource);
```
Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/graql/VarPattern.java]: resource?

File: grakn-core/src/main/java/ai/grakn/graql/VarPattern.java [hunk 7]
```diff
@@ -107,21 +107,33 @@
...
 
     /**
      * the variable must have a resource of the given type that matches the given atom
      *
      * @param type a resource type in the ontology
-     * @param varPattern a variable pattern representing a resource
+     * @param resource a variable pattern representing a resource
```
Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/graql/VarPattern.java]: resource?

File: grakn-core/src/main/java/ai/grakn/graql/VarPattern.java [hunk 8]
```diff
@@ -107,21 +107,33 @@
...
      * @return this
      */
     @CheckReturnValue
-    VarPattern has(Label type, VarPattern varPattern);
+    VarPattern has(Label type, VarPattern resource);
+
+    /**
+     * the variable must have a resource of the given type that matches {@code resource}. The relation associating
```
Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/graql/VarPattern.java]: resource? 
relation?

File: grakn-graph/src/main/java/ai/grakn/graph/internal/concept/RelationshipImpl.java [hunk 9]
```diff
@@ -95,11 +95,16 @@ public RelationshipStructure structure(){
 
     @Override
     public Relationship resource(Resource resource) {
-        reify().resource(resource);
+        resourceRelationship(resource);
         return this;
     }
 
     @Override
+    public Relationship resourceRelationship(Resource resource) {
```
Comment (fppt) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/concept/RelationshipImpl.java]: Why was this change needed ? Also resource?

File: grakn-graph/src/main/java/ai/grakn/graph/internal/concept/ThingImpl.java [hunk 10]
```diff
@@ -233,14 +233,19 @@ public String getIndex(){
...
      * @return The instance itself
      */
     @Override
     public T resource(Resource resource){
+        resourceRelationship(resource);
+        return getThis();
+    }
+
+    @Override
+    public Relationship resourceRelationship(Resource resource) {
```
Comment (fppt) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/concept/ThingImpl.java]: I think we are exposing a tad much now

File: grakn-graql/src/main/java/ai/grakn/graql/internal/pattern/AbstractVarPattern.java [hunk 11]
```diff
@@ -196,13 +196,18 @@ public final VarPattern has(String type, ValuePredicate predicate) {
     }
 
     @Override
-    public final VarPattern has(String type, VarPattern varPattern) {
-        return has(Label.of(type), varPattern);
+    public final VarPattern has(String type, VarPattern resource) {
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/pattern/AbstractVarPattern.java]: resource?

File: grakn-graql/src/main/java/ai/grakn/graql/internal/pattern/AbstractVarPattern.java [hunk 12]
```diff
@@ -196,13 +196,18 @@ public final VarPattern has(String type, ValuePredicate predicate) {
...
+        return has(Label.of(type), resource);
     }
 
     @Override
-    public final VarPattern has(Label type, VarPattern varPattern) {
-        return addProperty(HasResourceProperty.of(type, varPattern.admin()));
+    public final VarPattern has(Label type, VarPattern resource) {
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/pattern/AbstractVarPattern.java]: resource?

File: grakn-graql/src/main/java/ai/grakn/graql/internal/pattern/AbstractVarPattern.java [hunk 13]
```diff
@@ -196,13 +196,18 @@ public final VarPattern has(String type, ValuePredicate predicate) {
...
+        return has(type, resource, Graql.var());
+    }
+
+    @Override
+    public final VarPattern has(Label type, VarPattern resource, VarPattern relation) {
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/pattern/AbstractVarPattern.java]: resource?
relation?

File: grakn-graql/src/main/java/ai/grakn/graql/internal/pattern/property/HasResourceProperty.java [hunk 14]
```diff
@@ -73,14 +74,14 @@
 @AutoValue
 public abstract class HasResourceProperty extends AbstractVarProperty implements NamedProperty {
 
-    public static HasResourceProperty of(Label resourceType, VarPatternAdmin resource) {
+    public static HasResourceProperty of(Label resourceType, VarPatternAdmin resource, VarPatternAdmin relation) {
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/pattern/property/HasResourceProperty.java]: Oh screw this it's ⭐️ time

File: grakn-graql/src/main/java/ai/grakn/graql/internal/pattern/property/HasResourceProperty.java [hunk 15]
```diff
@@ -73,14 +74,14 @@
...
         resource = resource.isa(label(resourceType)).admin();
-        return new AutoValue_HasResourceProperty(resourceType, resource);
+        return new AutoValue_HasResourceProperty(resourceType, resource, relation);
     }
 
     public abstract Label type();
-
     public abstract VarPatternAdmin resource();
+    public abstract VarPatternAdmin relation();
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/pattern/property/HasResourceProperty.java]: ⭐️

File: grakn-graql/src/main/java/ai/grakn/graql/internal/pattern/property/HasResourceProperty.java [hunk 16]
```diff
@@ -166,8 +178,43 @@ private boolean testPredicate(Optional<ValuePredicateAdmin> optPredicate, Relati
         return Stream.of(label(type()).admin());
     }
 
+    private boolean hasReifiedRelation() {
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/pattern/property/HasResourceProperty.java]: ⭐️

File: grakn-graql/src/test/java/ai/grakn/graql/internal/parser/QueryParserTest.java [hunk 17]
```diff
@@ -789,6 +788,11 @@ public void whenParsingAVeryLargeQuery_DontRunOutOfMemory() {
         assertEquals(bigNumber, count[0]);
     }
 
+    @Test
+    public void whenParsingAQueryWithReifiedResourceRelationSyntax_ItIsEquivalentToJavaGraql() {
```
Comment (fppt) [inline grakn-graql/src/test/java/ai/grakn/graql/internal/parser/QueryParserTest.java]: ⭐️

File: grakn-graql/src/test/java/ai/grakn/graql/internal/parser/QueryParserTest.java [hunk 18]
```diff
@@ -789,6 +788,11 @@ public void whenParsingAVeryLargeQuery_DontRunOutOfMemory() {
...
+        assertParseEquivalence("match $x has name $z as $x;");
+    }
+
     @Test(expected = GraqlSyntaxException.class)
     public void testMultipleQueriesThrowsSyntaxException() {
```
Comment (fppt) [inline grakn-graql/src/test/java/ai/grakn/graql/internal/parser/QueryParserTest.java]: Nitpick: Crap Test Name

File: grakn-graql/src/test/java/ai/grakn/graql/internal/query/InsertQueryTest.java [hunk 19]
```diff
@@ -607,11 +619,46 @@ public void whenChangingTheSuperOfAnExistingConcept_ApplyTheChange() {
         EntityType movie = movieGraph.graph().getEntityType("movie");
 
         qb.match(var("x").label("a-new-type")).insert(var("x").sub("movie")).execute();
-        
+
         assertEquals(movie, newType.sup());
     }
 
     @Test
+    public void whenAddingAResourceRelationWithProvenance_TheResourceAndProvenanceAreAdded() {
```
Comment (fppt) [inline grakn-graql/src/test/java/ai/grakn/graql/internal/query/InsertQueryTest.java]: ⭐️

File: grakn-graql/src/test/java/ai/grakn/graql/internal/query/InsertQueryTest.java [hunk 20]
```diff
@@ -607,11 +619,46 @@ public void whenChangingTheSuperOfAnExistingConcept_ApplyTheChange() {
...
+        Resource<String> theTitle = answer.get(x).asResource();
+        Relationship hasTitle = answer.get(y).asRelationship();
+        Resource<String> provenance = answer.get(z).asResource();
+
+        assertThat(hasTitle.rolePlayers().toArray(), arrayContainingInAnyOrder(movie, theTitle));
+        assertThat(hasTitle.resources().toArray(), arrayContaining(provenance));
+    }
+
+    @Test
+    public void whenAddingProvenanceToAnExistingRelation_TheProvenanceIsAdded() {
```
Comment (fppt) [inline grakn-graql/src/test/java/ai/grakn/graql/internal/query/InsertQueryTest.java]: ⭐️

File: grakn-graql/src/test/java/ai/grakn/graql/internal/query/QueryBuilderTest.java [hunk 21]
```diff
@@ -56,19 +60,19 @@ public void clear(){
...
     @Test
     public void testBuildMatchQueryGraphLast() {
-        MatchQuery query = match(var("x").isa("movie")).withGraph(movieGraph.graph());
-        assertThat(query, variable("x", containsAllMovies));
+        MatchQuery query = match(x.isa("movie")).withGraph(movieGraph.graph());
+        assertThat(query, variable(x, containsAllMovies));
     }
 
     @Test
     public void testBuildAskQueryGraphLast() {
```
Comment (fppt) [inline grakn-graql/src/test/java/ai/grakn/graql/internal/query/QueryBuilderTest.java]: Test Name and a bunch of other test names . . .

File: grakn-graql/src/test/java/ai/grakn/graql/internal/query/match/MatchQueryModifierTest.java [hunk 22]
```diff
@@ -65,116 +70,119 @@ public void setUp() {
...
+        Var t = var("t");
+        MatchQuery query = qb.match(x.isa("movie").has("title", t)).orderBy(t, asc).offset(1).limit(3);
 
-        assertResultsOrderedByValue(query, "t", true);
+        assertResultsOrderedByValue(query, t, true);
         assertEquals(3, query.stream().count());
     }
 
     @Test
     public void testOrPatternOrderByResource() {
```
Comment (fppt) [inline grakn-graql/src/test/java/ai/grakn/graql/internal/query/match/MatchQueryModifierTest.java]: mor bad test names

File: grakn-core/src/main/java/ai/grakn/concept/Thing.java [hunk 23]
```diff
@@ -80,6 +80,14 @@
...
+     *
+     * @param resource The resource to which a relationship is created
+     * @return The relationship connecting the thing and the resource
+     */
+    Relationship resourceRelationship(Resource resource);
```
Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/concept/Thing.java]: Is there no way to not expose this?

Comment (aelred) [inline grakn-core/src/main/java/ai/grakn/concept/Thing.java]: I feel like if Graql can do it, then you should be able to do it with the graph API too. Maybe there's a better way to present it though.

File: grakn-graph/src/main/java/ai/grakn/graph/internal/concept/RelationshipImpl.java (hunk 9 above)
Comment (aelred) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/concept/RelationshipImpl.java]: Here I'm defining `resource` in terms of `resourceRelationship` so I don't repeat myself.

**Comment by aelred** [comment]
@fppt re-review time :D

**Comment by aelred** [comment]
@fppt we've added the reasoner changes, plus a big merge with master so you'll want to review again
